visual-dotnet/SauceLabs.Visual/Utils/DictionaryExtensions.cs

```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SauceLabs.Visual.Utils
+{
+    internal static class DictionaryExtensions
+    {
+        private static readonly SemaphoreSlim Semaphore = new SemaphoreSlim(1, 1);
+
+        public static async Task<TValue> GetOrAddAsync<TKey, TValue>(this IDictionary<TKey, TValue> dict, TKey key, Func<TKey, Task<TValue>> valueProvider)
+        {
+            await Semaphore.WaitAsync();
+            try
+            {
+                if (dict == null) throw new ArgumentNullException(nameof(dict));
+                if (key == null) throw new ArgumentNullException(nameof(key));
+                if (valueProvider == null) throw new ArgumentNullException(nameof(valueProvider));
+
```

Comment:
small nit:
in the project we're adding enter after each if statement + we're using brackets. Should we also change it?